### PR TITLE
Added `Generate Recon Config` support to `databricks labs remorph generate-recon-config` cli command

### PR DESCRIPTION
**Pull request by vijaypavann-db**
## Changes

Added CLI command `databricks labs remorph generate-recon-config`  to generate Recon Config for Data Reconciliation.

Logic:
-  This command prompts user for the following: 
      * Source
      * Secret Scope
      * Source, Target Catalog and Schema
      * include/exclude a set of tables
- Based on the provided information it will generate `recon-config-file`
- This feature is intended to enhance our tool's capabilities in generating the `recon-config-file` dynamically which will eventually minimise the unintended manual errors and streamline the process of prompting and saving configuration details. 

- Sample `recon-config-file`

```
{
  "source_catalog": "catalog",
  "source_schema": "schema",
  "tables": [
    {
      "column_mapping": None,
      "drop_columns": None,
      "filters": None,
      "jdbc_reader_options": None,
      "join_columns": None,
      "select_columns": None,
      "source_name": "table1",
      "target_name": "table1",
      "thresholds": None,
      "transformations": None
    },
    {
      "column_mapping": None,
      "drop_columns": None,
      "filters": None,
      "jdbc_reader_options": None,
      "join_columns": None,
      "select_columns": None,
      "source_name": "table2",
      "target_name": "table2",
      "thresholds": None,
      "transformations": None
    },
    {
      "column_mapping": None,
      "drop_columns": None,
      "filters": None,
      "jdbc_reader_options": None,
      "join_columns": None,
      "select_columns": None,
      "source_name": "table3",
      "target_name": "table3",
      "thresholds": None,
      "transformations": None
    }
  ],
  "target_catalog": "tgt_catalog",
  "target_schema": "tgt_schema"
}
```

### Linked issues
Resolves 
https://github.com/databrickslabs/remorph/issues/209

### Functionality
- [x]  added new CLI command

### Tests
- [x]  manually tested
- [x]  added unit tests

### TODO

- [ ]  add integration tests
- [ ]  verify on staging environment (attach screenshot )

### CODE REVIEW

File: src/databricks/labs/remorph/reconcile/connectors/snowflake.py [hunk 1]
```diff
@@ -1,12 +1,21 @@
...
 from databricks.labs.remorph.reconcile.constants import SourceDriver, SourceType
-from databricks.labs.remorph.reconcile.recon_config import JdbcReaderOptions, Schema
+from databricks.labs.remorph.reconcile.recon_config import (
+    JdbcReaderOptions,
+    Schema,
+    Table,
+    TableRecon,
+)
+
+logger = get_logger(__file__)
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/reconcile/connectors/snowflake.py:18]: same as above

File: src/databricks/labs/remorph/reconcile/connectors/data_source.py [hunk 2]
```diff
@@ -25,13 +30,23 @@ def __init__(
...
 
+    @abstractmethod
+    def list_tables(
+        self,
+        catalog: str,
+        schema: str,
+        include_list: list[str] | None,
+        exclude_list: list[str] | None,
+    ) -> TableRecon:
+        return NotImplemented
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/reconcile/connectors/data_source.py:48]: is this expected to be covered in a different PR?

**Review by sundarshankar89 — CHANGES_REQUESTED** [review]

File: src/databricks/labs/remorph/reconcile/connectors/data_source.py [hunk 3]
```diff
@@ -1,14 +1,19 @@
...
 
-from databricks.labs.remorph.reconcile.recon_config import (  # pylint: disable=ungrouped-imports
+# pylint: disable-next=ungrouped-imports
+from databricks.labs.remorph.reconcile.recon_config import (
     JdbcReaderOptions,
     Schema,
+    TableRecon,
 )
 
+logger = get_logger(__file__)
```
[the remark this answers is not in the document]
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/reconcile/connectors/data_source.py:15]: Done

File: src/databricks/labs/remorph/reconcile/connectors/data_source_factory.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+from pyspark.sql import SparkSession
+
+# pylint: disable-next=ungrouped-imports
+from databricks.labs.remorph.reconcile.connectors.data_source import DataSource
+from databricks.labs.remorph.reconcile.connectors.databricks import DatabricksDataSource
+from databricks.labs.remorph.reconcile.connectors.oracle import OracleDataSource
+from databricks.labs.remorph.reconcile.connectors.snowflake import SnowflakeDataSource
+from databricks.labs.remorph.reconcile.constants import SourceType
+
+logger = get_logger(__file__)
```
[the remark this answers is not in the document]
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/reconcile/connectors/data_source_factory.py:12]: Done

File: src/databricks/labs/remorph/reconcile/connectors/snowflake.py (hunk 1 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/reconcile/connectors/snowflake.py:18]: Done

File: src/databricks/labs/remorph/reconcile/connectors/data_source.py (hunk 2 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/reconcile/connectors/data_source.py:48]: Yes, will be 2 different PRs for Oracle and Snowflake

File: src/databricks/labs/remorph/helpers/db_workspace_utils.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
+import logging
+
+from databricks.labs.blueprint.tui import Prompts
+from databricks.sdk import WorkspaceClient
+
+logger = logging.getLogger(__name__)
+
+
+class DatabricksSecretsClient:
```
Comment (nfx) [inline src/databricks/labs/remorph/helpers/db_workspace_utils.py:9]: it looks like it's redundant because of https://github.com/databrickslabs/remorph/pull/254

File: src/databricks/labs/remorph/reconcile/connectors/data_source_factory.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
+import logging
+
+from databricks.sdk import WorkspaceClient  # pylint: disable-next=wrong-import-order
+from pyspark.sql import SparkSession
+
+# pylint: disable-next=ungrouped-imports
```
Comment (nfx) [inline src/databricks/labs/remorph/reconcile/connectors/data_source_factory.py:6]: don't disable pylint!

File: src/databricks/labs/remorph/reconcile/connectors/data_source_factory.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+from databricks.labs.remorph.reconcile.connectors.oracle import OracleDataSource
+from databricks.labs.remorph.reconcile.connectors.snowflake import SnowflakeDataSource
+from databricks.labs.remorph.reconcile.constants import SourceType
+
+logger = logging.getLogger(__name__)
+
+
+class DataSourceFactory:
+    @staticmethod
+    def get_data_source(
```
Comment (nfx) [inline src/databricks/labs/remorph/reconcile/connectors/data_source_factory.py:18]: this method is redundant here. move it as private method to `ReconConfigPrompts`

File: src/databricks/labs/remorph/helpers/recon_config_utils.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+        return recon_config
+
+    def _save_config_details(self, recon_config_json):
+        """
+        Save the config details in a file
+        """
+        recon_conf_abspath = os.path.abspath(f"./recon_conf_{self._source}.json")
+        logger.debug(f"Saving the config details for `{self._source}` in `{recon_conf_abspath}` file")
+        with open(f"./recon_conf_{self._source}.json", "w", encoding="utf-8") as f:
```
Comment (nfx) [inline src/databricks/labs/remorph/helpers/recon_config_utils.py:124]: use `Installation`, don't save on local fs:
https://github.com/databrickslabs/blueprint?tab=readme-ov-file#saving-dataclass-configuration

File: src/databricks/labs/remorph/reconcile/connectors/data_source.py [hunk 9]
```diff
@@ -1,14 +1,19 @@
+import logging
 import re
 from abc import ABC, abstractmethod
 
 from databricks.sdk import WorkspaceClient  # pylint: disable-next=wrong-import-order
 from pyspark.sql import DataFrame, SparkSession
 
-from databricks.labs.remorph.reconcile.recon_config import (  # pylint: disable=ungrouped-imports
+# pylint: disable-next=ungrouped-imports
```
Comment (nfx) [inline src/databricks/labs/remorph/reconcile/connectors/data_source.py:8]: don't disable pylint!

File: tests/unit/helpers/test_db_workspace_utils.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
+# pylint: disable=protected-access
```
Comment (nfx) [inline tests/unit/helpers/test_db_workspace_utils.py:1]: DO NOT DISABLE PYLINT

**Review by nfx — CHANGES_REQUESTED** [review]
this PR is incorrect and has bugs

File: src/databricks/labs/remorph/reconcile/connectors/data_source.py (hunk 9 above)
Comment (sundarshankar89) [reply src/databricks/labs/remorph/reconcile/connectors/data_source.py:8]: isort is causing issue here, we will need to allow for `ungrouped-imports` `wrong-import-order` and this will be fixed in a separate PR

File: src/databricks/labs/remorph/reconcile/connectors/data_source_factory.py (hunk 7 above)
Comment (sundarshankar89) [reply src/databricks/labs/remorph/reconcile/connectors/data_source_factory.py:18]: this needs to be inside config.py/lib.py as a method which calls private dict which has this mapping and we do dict lookup check https://github.com/databrickslabs/remorph/blob/main/src/databricks/labs/remorph/config.py#L10

File: tests/unit/helpers/test_db_workspace_utils.py (hunk 10 above)
Comment (sundarshankar89) [reply tests/unit/helpers/test_db_workspace_utils.py:1]: Yes private method UT will happen by calling the parent method, no need to test private methods directly.

**Review by sundarshankar89 — CHANGES_REQUESTED** [review]

File: src/databricks/labs/remorph/helpers/db_workspace_utils.py (hunk 5 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/helpers/db_workspace_utils.py:9]: Refactored the `DatabricksSecretsClient` to `ReconConfigPrompts`, few functions are used in both the PR's. 
That could be the reason it's looking as redundant, but the PR's are segregated according to the functionality.

File: src/databricks/labs/remorph/reconcile/connectors/data_source_factory.py (hunk 7 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/reconcile/connectors/data_source_factory.py:18]: moved to `config.py` but dict lookup won't be possible, because we need to initialize the class with Spark and WorkspaceClient

File: src/databricks/labs/remorph/reconcile/connectors/data_source.py (hunk 9 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/reconcile/connectors/data_source.py:8]: Handled in #270

File: tests/unit/helpers/test_db_workspace_utils.py (hunk 10 above)
Comment (vijaypavann-db) [reply tests/unit/helpers/test_db_workspace_utils.py:1]: Done

**Comment by vijaypavann-db** [comment]
* Installation.save :  `ws_file_url = installation.save(recon_config, filename=recon_conf_file)`
     - `Installation.save` generates the following config file with **no default values** and without any **keys sorting**:
     
     ```
     {
  "source_schema": "catalog",
  "target_catalog": "tgt_catalog",
  "target_schema": "tgt_schema",
  "tables": [
    {
      "source_name": "part",
      "target_name": "part"
    },
    {
      "source_name": "orders",
      "target_name": "orders"
    }
  ],
  "source_catalog": "snowflake_sample_data",
  "version": 1
    }     
```
Where as
*  Installation.upload :
recon_config_json = (json.dumps(recon_config, default=vars, indent=2, sort_keys=True)
                             .replace("null", "None"))
installation.upload(filename=recon_conf_file, raw=recon_config_json.encode("utf-8")) 
        - Using Installation.upload we can generate the config file with the **default values** and **sorted keys** as below :

{
  "source_catalog": "catalog",
  "source_schema": "schema",
  "tables": [
    {
      "column_mapping": None,
      "drop_columns": None,
      "filters": None,
      "jdbc_reader_options": None,
      "join_columns": None,
      "select_columns": None,
      "source_name": "table1",
      "target_name": "table1",
      "thresholds": None,
      "transformations": None
    },
    {
      "column_mapping": None,
      "drop_columns": None,
      "filters": None,
      "jdbc_reader_options": None,
      "join_columns": None,
      "select_columns": None,
      "source_name": "table2",
      "target_name": "table2",
      "thresholds": None,
      "transformations": None
    },
    {
      "column_mapping": None,
      "drop_columns": None,
      "filters": None,
      "jdbc_reader_options": None,
      "join_columns": None,
      "select_columns": None,
      "source_name": "table3",
      "target_name": "table3",
      "thresholds": None,
      "transformations": None
    }
  ],
  "target_catalog": "tgt_catalog",
  "target_schema": "tgt_schema"
}

```
As per our design, we should generate a (readable) Template config file with all the available options for the selected tables for the given source Catalog and Schema.  Hence, using `upload` instead of `save`.

File: src/databricks/labs/remorph/reconcile/connectors/data_source_factory.py (hunk 7 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/reconcile/connectors/data_source_factory.py]: Can't refactor `get_data_source` into `config.py` due to the `circular import` error, moved to `recon_config_utils.py`:

```
 File "remorph/src/databricks/labs/remorph/config.py", line 8, in <module>
    from databricks.labs.remorph.reconcile.connectors.data_source import DataSource
  File "remorph/src/databricks/labs/remorph/reconcile/connectors/data_source.py", line 7, in <module>
    from databricks.labs.remorph.config import TableRecon
ImportError: cannot import name 'TableRecon' from partially initialized module 'databricks.labs.remorph.config' (most likely due to a circular import) 
(remorph/src/databricks/labs/remorph/config.py)
Error: exit status 1
10:10:34 ERROR failed execution pid=68014 exit_code=1 error="exit status 1"
```

File: src/databricks/labs/remorph/cli.py [hunk 11]
```diff
@@ -89,6 +89,18 @@ def reconcile(w: WorkspaceClient, recon_conf: str, conn_profile: str, source: st
...
+    recon_conf = ReconConfigPrompts(w)
+
+    # Prompt for source
+    recon_conf.prompt_source()
+
+    # Prompt for connection details and save the config
+    recon_conf.prompt_and_save_config_details()
+
+
 def generate_lineage(w: WorkspaceClient, source: str, input_sql: str, output_folder: str):
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/cli.py:104]: some merge conflict here please fix `@remorph.command` missing before `generate_lineage`

File: src/databricks/labs/remorph/helpers/recon_config_utils.py [hunk 12]
```diff
@@ -14,6 +29,24 @@
 ]
 
 
+def get_data_source(
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/helpers/recon_config_utils.py:32]: move to the connectors `client.py`

**Review by sundarshankar89 — CHANGES_REQUESTED** [review]

File: src/databricks/labs/remorph/cli.py (hunk 11 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/cli.py:105]: Done, added missing statement

File: src/databricks/labs/remorph/helpers/recon_config_utils.py (hunk 12 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/helpers/recon_config_utils.py]: Done

File: src/databricks/labs/remorph/helpers/recon_config_utils.py [hunk 13]
```diff
@@ -1,10 +1,20 @@
...
 from databricks.labs.blueprint.tui import Prompts
+from databricks.labs.blueprint.wheels import ProductInfo
+from databricks.labs.remorph.__about__ import __version__
+from databricks.labs.remorph.config import TableRecon
+from databricks.labs.remorph.reconcile.connectors.client import get_data_source
 from databricks.labs.remorph.reconcile.constants import SourceType
 from databricks.sdk import WorkspaceClient
 from databricks.sdk.errors.platform import ResourceDoesNotExist
 
+PRODUCT_INFO = ProductInfo(__file__)
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/helpers/recon_config_utils.py:16]: Why do you needs this? `WorkspaceClient` should already have this variable.

**Review by sundarshankar89 — APPROVED** [review]
LGTM

File: src/databricks/labs/remorph/helpers/recon_config_utils.py (hunk 13 above)
Comment (sundarshankar89) [reply src/databricks/labs/remorph/helpers/recon_config_utils.py:16]: @vijaypavann-db is this fixed?

Comment (vijaypavann-db) [reply src/databricks/labs/remorph/helpers/recon_config_utils.py]: Done Sundar,  Hardcoded `product as remorph` for now, as `WorkspaceClient`  `product` variable is [not exposed](https://databricks-sdk-py.readthedocs.io/en/latest/clients/workspace.html)

**Review by sundarshankar89 — DISMISSED** [review]
LGTM

File: src/databricks/labs/remorph/helpers/recon_config_utils.py [hunk 14]
```diff
@@ -98,6 +104,117 @@ def prompt_source(self):
...
+    def prompt_and_save_config_details(self):
+        """
+        Prompt for source, target catalog and schema names. Get the table list and save the config details
+        """
+        # Check for Secrets Scope
+        self._confirm_secret_scope()
+        recon_config, recon_config_file = self._prompt_config_details()
+        recon_config_json = json.dumps(recon_config, default=vars, indent=2, sort_keys=True)
+        logger.debug(f"recon_config_json : {recon_config_json}")
+        self._save_config_details(recon_config_json, recon_config_file)
```
Comment (nfx) [inline src/databricks/labs/remorph/helpers/recon_config_utils.py:216]:
```suggestion
        self._installation.save(recon_config, filename=recon_config_file)
        if self._prompts.confirm(f"Open config file in the browser?"):
            workspace_url = self._installation.workspace_link(recon_config_file)
            webbrowser.open(workspace_url)
```

**please don't re-invent the wheel.** Also, inject `self._installation` via constructor, so that you can use the `MockInstallation`: https://github.com/databrickslabs/blueprint?tab=readme-ov-file#unit-testing-installation-state

File: src/databricks/labs/remorph/reconcile/connectors/databricks.py [hunk 15]
```diff
@@ -4,12 +4,18 @@
...
+from databricks.labs.remorph.config import TableRecon
 from databricks.labs.remorph.reconcile.connectors.data_source import DataSource
+from databricks.labs.remorph.reconcile.constants import SourceType
 from databricks.labs.remorph.reconcile.recon_config import JdbcReaderOptions, Schema
 
 
 class DatabricksDataSource(DataSource):
-    def read_data(self, catalog: str, schema: str, query: str, options: JdbcReaderOptions) -> DataFrame:
+    def __init__(self, spark, ws, scope):
+        super().__init__(spark, ws, scope, engine=SourceType.DATABRICKS.value)
```
Comment (nfx) [inline src/databricks/labs/remorph/reconcile/connectors/databricks.py:15]: not blocking this PR, but it's code-smell, because you're not injecting the instance of DatabricksDialect and use an enum instead.

File: src/databricks/labs/remorph/reconcile/connectors/data_source.py [hunk 16]
```diff
@@ -1,34 +1,48 @@
...
 
 class DataSource(ABC):
     # TODO need to remove connection_params
     def __init__(
         self,
-        engine: str,
         spark: SparkSession,
         ws: WorkspaceClient,
         scope: str,
+        engine: str = None,
```
Comment (nfx) [inline src/databricks/labs/remorph/reconcile/connectors/data_source.py:21]:
```suggestion
        engine: Dialect,
```

not blocking this PR, but it's a code smell when you have to be passing down the instance of `sqlglot.dialects.dialect.Dialect` and instead pass a string to reference it. this will blow up very soon.

also, you have `SparkSession` and `WorkspaceClient` as "global object instances" passed via their instances, so be consistent with that.

File: tests/unit/helpers/test_recon_config_utils.py [hunk 17]
```diff
@@ -143,3 +153,112 @@ def test_store_connection_secrets_overwrite(mock_workspace_client):
...
+        target_schema="tgt_schema",
+        tables=[Table(source_name="table1", target_name="table1"), Table(source_name="table2", target_name="table2")],
+    )
+
+    # Patch the scope_exists method to return True
+    mock_workspace_client.secrets.list_scopes.side_effect = [[SecretScope(name="dummy")]]
+    # Patch the builder method to return a SparkSession
+    with patch.object(DatabricksSession, "builder", MagicMock(return_value=SparkSession.builder)):
+        # mock WorkspaceClient and Installation in _save_config_details
+        with patch("databricks.labs.remorph.helpers.recon_config_utils.WorkspaceClient", mock_workspace_client):
```
Comment (nfx) [inline tests/unit/helpers/test_recon_config_utils.py:230]: Obscure implicit test dependency with mock.patch(`databricks....`). Rewrite to inject dependencies through constructor.. Using patch to mock dependencies in unit tests can introduce implicit dependencies within a class, making it unclear to other developers. Constructor arguments, on the other hand, explicitly declare dependencies, enhancing code readability and maintainability. However, reliance on patch for testing may lead to issues during refactoring, as updates to underlying implementations would necessitate changes across multiple unrelated unit tests. Moreover, the use of hard-coded strings in patch can obscure which unit tests require modification, as they lack strongly typed references. This coupling of the class under test to concrete classes signifies a code smell, and such code is not easily portable to statically typed languages where monkey patching isn't feasible without significant effort. In essence, extensive patching of external clients suggests a need for refactoring, with experienced engineers recognizing the potential for dependency inversion in such scenarios.

To address this issue, refactor the code to inject dependencies through the constructor. This approach explicitly declares dependencies, enhancing code readability and maintainability. Moreover, it allows for dependency inversion, enabling the use of interfaces to decouple the class under test from concrete classes. This decoupling facilitates unit testing, as it allows for the substitution of mock objects for concrete implementations, ensuring that the class under test behaves as expected. By following this approach, you can create more robust and maintainable unit tests, improving the overall quality of your codebase.

https://github.com/databrickslabs/pylint-plugin?tab=readme-ov-file#r8918-explicit-dependency-required

File: tests/unit/helpers/test_recon_config_utils.py [hunk 18]
```diff
@@ -143,3 +153,112 @@ def test_store_connection_secrets_overwrite(mock_workspace_client):
...
+            with patch("databricks.labs.remorph.helpers.recon_config_utils.Installation", MagicMock()):
```
Comment (nfx) [inline tests/unit/helpers/test_recon_config_utils.py:231]: use `MockInstallation` - https://github.com/databrickslabs/blueprint?tab=readme-ov-file#unit-testing-installation-state

File: tests/unit/test_cli.py [hunk 19]
```diff
@@ -305,6 +305,12 @@ def test_recon_with_valid_input(mock_workspace_client_cli):
         mock_recon.assert_called_once_with(recon_conf, conn_profile, source, report)
 
 
+def test_cli_generate_recon_config(mock_workspace_client):
+    with patch("databricks.labs.remorph.cli.ReconConfigPrompts") as mock_recon_config:
```
Comment (nfx) [inline tests/unit/test_cli.py:309]: see https://github.com/databrickslabs/pylint-plugin?tab=readme-ov-file#r8918-explicit-dependency-required

File: src/databricks/labs/remorph/reconcile/connectors/client.py [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+from databricks.labs.remorph.reconcile.connectors.databricks import DatabricksDataSource
+from databricks.labs.remorph.reconcile.connectors.oracle import OracleDataSource
+from databricks.labs.remorph.reconcile.connectors.snowflake import SnowflakeDataSource
+from databricks.labs.remorph.reconcile.constants import SourceType
+from databricks.sdk import WorkspaceClient
+
+logger = logging.getLogger(__name__)
+
+
+def get_data_source(
```
Comment (nfx) [inline src/databricks/labs/remorph/reconcile/connectors/client.py:15]: see the comment on `DataSource`

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/remorph/helpers/recon_config_utils.py (hunk 14 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/helpers/recon_config_utils.py:216]: Sure, will inject `self._installation` via constructor. 
But we couldn't use `installation.save` due to the reasons mentioned [here](https://github.com/databrickslabs/remorph/pull/249#issuecomment-2080688538)

File: src/databricks/labs/remorph/cli.py [hunk 21]
```diff
@@ -96,6 +96,21 @@ def _get_spark_session(ws: WorkspaceClient) -> SparkSession:
...
+@remorph.command
+def generate_recon_config(w: WorkspaceClient):
+    """generates recon_config file for reconciliation"""
+    logger.info(f"User: {w.current_user.me()}")
+    logger.info("Generating recon_config file for reconcile")
+
+    # Create Installation object
+    # installation = Installation(w, "remorph")
+    folder_path = f"/Users/{w.current_user.me().user_name}/.remorph_demo"
+    installation = Installation(w, "remorph", install_folder=folder_path)
```
Comment (nfx) [inline src/databricks/labs/remorph/cli.py:108]:
```suggestion
    installation = Installation.assume_user_home(ws, "remorph")
```

https://github.com/databrickslabs/blueprint?tab=readme-ov-file#detecting-current-installation

File: src/databricks/labs/remorph/helpers/reconcile_config_utils.py [hunk 22]
```diff
@@ -0,0 +1,390 @@
...
+# MAGIC %md
+# MAGIC # Recon Config setup instructions (see [README](readme_link))
+# MAGIC
+# MAGIC Production runs are supposed to be triggered through the following jobs: job_links
+# MAGIC
+# MAGIC **This notebook is overwritten with each `remorph` generate-recon-config.**
+
+# COMMAND ----------
+
+# MAGIC %pip install remorph_VERSION-py3-none-any.whl
```
Comment (nfx) [inline src/databricks/labs/remorph/helpers/reconcile_config_utils.py:26]: upload wheel to WSFS and replace the link in here - see how to do it in https://github.com/databrickslabs/blueprint?tab=readme-ov-file#publishing-wheels-to-databricks-workspace

File: src/databricks/labs/remorph/helpers/reconcile_config_utils.py [hunk 23]
```diff
@@ -0,0 +1,390 @@
...
+from databricks.labs.remorph.reconcile.connectors.source_adapter import create_adapter
+from databricks.labs.remorph.reconcile.constants import SourceType
+from databricks.sdk import WorkspaceClient
+from databricks.sdk.errors.platform import ResourceDoesNotExist, NotFound, PermissionDenied
+
+logger = logging.getLogger(__name__)
+
+README_RECON_CONFIG_NOTEBOOK = """# Databricks notebook source
+# MAGIC %md
+# MAGIC # Recon Config setup instructions (see [README](readme_link))
```
Comment (nfx) [inline src/databricks/labs/remorph/helpers/reconcile_config_utils.py:18]: don't you want to replace `readme_link` with a real link?

File: src/databricks/labs/remorph/helpers/reconcile_config_utils.py [hunk 24]
```diff
@@ -0,0 +1,390 @@
...
+"""
+
+
+class ReconcileConfigUtils:
+    def __init__(self, ws: WorkspaceClient, installation: Installation = None, prompts: Prompts = Prompts()):
+        self._source = None
+        self._prompts = prompts
+        self._ws = ws
+        self._installation = installation
+        self._ws_reconcile_config: ReconcileConfig | None = None
```
Comment (nfx) [inline src/databricks/labs/remorph/helpers/reconcile_config_utils.py:63]: a bit of a nitpick: we already have `self._ws` so this way it'll be less confusing:
```suggestion
        self._reconcile_config: ReconcileConfig | None = None
```

File: src/databricks/labs/remorph/helpers/reconcile_config_utils.py [hunk 25]
```diff
@@ -0,0 +1,390 @@
...
+    tgt_schema,
+    get_dialect("snowflake"),
+    table_conf,
+)
+df = schema_compare_output.compare_df
+"""
+
+
+class ReconcileConfigUtils:
+    def __init__(self, ws: WorkspaceClient, installation: Installation = None, prompts: Prompts = Prompts()):
```
Comment (nfx) [inline src/databricks/labs/remorph/helpers/reconcile_config_utils.py:58]:
```suggestion
    def __init__(self, ws: WorkspaceClient, installation: Installation, prompts: Prompts = Prompts()):
```
don't make things optional when they are required.

File: src/databricks/labs/remorph/helpers/reconcile_config_utils.py [hunk 26]
```diff
@@ -0,0 +1,390 @@
...
+        if not scope_exists:
+            allow_scope_creation = self._prompts.confirm("Do you want to create a new one?")
+            if not allow_scope_creation:
+                msg = f" `{scope_name}` Scope is needed to store Secrets in Databricks Workspace"
+                raise SystemExit(msg)
+
+            try:
+                logger.debug(f" Creating a new Scope: `{scope_name}`")
+                self._ws.secrets.create_scope(scope_name)
+            except Exception as ex:
```
Comment (nfx) [inline src/databricks/labs/remorph/helpers/reconcile_config_utils.py:92]: @sundarshankar89  why it's not caught by pylint? catch narrower exception.

File: src/databricks/labs/remorph/helpers/reconcile_config_utils.py [hunk 27]
```diff
@@ -0,0 +1,390 @@
...
+            return True
+        except ResourceDoesNotExist:
+            logger.debug(f"Secret key `{secret_key}` not found in Scope `{scope_name}`")
+            return False
+
+    def _store_secret(self, scope_name: str, secret_key: str, secret_value: str):
+        try:
+            logger.debug(f"Storing Secret: *{secret_key}* in Scope: `{scope_name}`")
+            self._ws.secrets.put_secret(scope=scope_name, key=secret_key, string_value=secret_value)
+        except Exception as ex:
```
Comment (nfx) [inline src/databricks/labs/remorph/helpers/reconcile_config_utils.py:112]: @sundarshankar89 why it's not caught by pylint? catch narrower exception.

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/remorph/helpers/reconcile_config_utils.py (hunk 24 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/helpers/reconcile_config_utils.py:63]: Done

File: src/databricks/labs/remorph/helpers/reconcile_config_utils.py (hunk 23 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/helpers/reconcile_config_utils.py:18]: this is just a template for now, we'll add more detailed examples and links

File: src/databricks/labs/remorph/helpers/reconcile_config_utils.py (hunk 22 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/helpers/reconcile_config_utils.py:26]: template added just for reference, we'll add more details in next PR

File: src/databricks/labs/remorph/cli.py (hunk 21 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/cli.py]: Done

**Comment by sundarshankar89** [comment]
Closing this and will be revisited
